### PR TITLE
#167377552 Pagination support for articles

### PR DESCRIPTION
**Pull request by Jonathan4github**
#### What does this PR do?
This pull request implement pagination support for articles.
#### Description of Task to be completed
- implement pagination functionality in article route
- implement validation for page and limit query parameter
#### How should this be manually tested?
- git clone this branch && cd to project
- run `npm install` && run `npm run test`
#### Any Background Context You want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#167377552](https://pivotaltracker.com/story/show/167377552)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

### CODE REVIEW

**Review by Merkll — COMMENTED** [review]
Make your pagination code reuseable

**Review by Merkll — APPROVED** [review]
NIce work modularizing the paginator

**Review by ibkadeeko — APPROVED** [review]

**Comment by iverenshaguy** [comment]
@Jonathan4github Please rebase on `develop` and re-push taking into consideration my instructions for the `.env` file format.

File: src/controllers/article.js [hunk 1]
```diff
@@ -22,5 +23,28 @@ export default {
...
+  searchArticles: async (request, response) => {
+    const { page = 1, limit = 10 } = request.query;
+    const { data: articles, count } = await paginator(Articles, { page, limit });
+
+    return response.status(200).json({
+      status: 'success',
+      data: {
+        count, page, current: +page, articles
+      },
+      message: 'Articles retrieve successfully'
```
Comment (iverenshaguy) [inline src/controllers/article.js]:
```suggestion
      message: 'Articles retrieved successfully'
```

File: src/helpers/paginator.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * @param {Object} options
+ *
+ * @returns {Object} returns object
+ */
+const paginator = async (Source, options) => {
+  const countResult = await Source.findAndCountAll({});
+  const { page, limit, ...otherOptions } = options;
+  const { count } = countResult;
+
+  if (!count) throw new NotFoundError('no articles found');
```
Comment (iverenshaguy) [inline src/helpers/paginator.js]:
```suggestion
  if (!count) throw new NotFoundError('No articles found');
```

File: src/helpers/paginator.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+  const pages = Math.ceil(count / limit);
+  const offset = limit * (+page - 1);
+
+  if (page > pages) throw new ApplicationError(400, 'page does not exist');
```
Comment (iverenshaguy) [inline src/helpers/paginator.js]:
```suggestion
  if (page > pages) throw new NotFoundError('Page does not exist');
```

File: src/validations/pagination.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { check } from 'express-validator/check';
+
+export default {
+  createPaginationSchema: [
+    check('page')
+      .optional()
+      .trim()
+      .isInt()
+      .withMessage('page must be a number')
```
Comment (iverenshaguy) [inline src/validations/pagination.js]:
```suggestion
      .withMessage('Page must be a number')
```

File: src/validations/pagination.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+      .customSanitizer(page => page.toLowerCase()),
+    check('limit')
+      .optional()
+      .trim()
+      .isNumeric()
+      .withMessage('limit must be a number')
```
Comment (iverenshaguy) [inline src/validations/pagination.js]:
```suggestion
      .withMessage('Limit must be a number')
```

**Review by iverenshaguy — APPROVED** [review]
Good job, please quickly update according to my comments and recent instructions of the order of the `.env` file

**Comment by iverenshaguy** [comment]
Your PR description has no response body screenshot. Please try not to repeat this is subsequent PRs

File: src/controllers/article.js [hunk 6]
```diff
@@ -22,5 +23,28 @@ export default {
...
+   * @return {Object} - callback that execute the controller
+   */
+  searchArticles: async (request, response) => {
+    const { page = 1, limit = 10 } = request.query;
+    const { data: articles, count } = await paginator(Articles, { page, limit });
+
+    return response.status(200).json({
+      status: 'success',
+      data: {
+        count, page, current: +page, articles
```
Comment (iverenshaguy) [inline src/controllers/article.js:45]: No need for current, please remove it in a subsequent PR. Thank you

**Pull request merged**